### PR TITLE
Stax port

### PR DESCRIPTION
**Pull request by agrojean-ledger**

### CODE REVIEW

File: src/ui/ui_menu_nbgl.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                         ui_menu_main,
+                         settingsNavCallback,
+                         settingsControlsCallback);
+}
+
+void ui_menu_main(void) {
+    switches[SWITCH_IDX(SWITCH_HASH_SET_TOKEN)].initState =
+        (HAS_SETTING(S_HASH_SIGNING_ENABLED)) ? ON_STATE : OFF_STATE;
+    switches[SWITCH_IDX(SWITCH_SEQUENCE_SET_TOKEN)].initState =
+        (HAS_SETTING(S_SEQUENCE_NUMBER_ENABLED)) ? ON_STATE : OFF_STATE;
```
Comment (xchapron-ledger) [inline src/ui/ui_menu_nbgl.c:102]: Can be moved to `settingsNavCallback` and then also removed from `settingsControlsCallback`

File: src/ui/ui_transaction_hash_nbgl.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    if (!format_hex(G_context.hash, 32, str_values[1], DETAIL_VALUE_MAX_LENGTH)) {
+        io_send_sw(SW_DISPLAY_TRANSACTION_HASH_FAIL);
+        return;
+    }
+    caption_value_pairs[1].value = str_values[1];
+}
+
+static void rejectConfirmation(void) {
+    ui_action_validate_transaction(false);
+    nbgl_useCaseStatus("TRANSACTION\nREJECTED", false, ui_menu_main);
```
Comment (xchapron-ledger) [inline src/ui/ui_transaction_hash_nbgl.c:72]: Should not be full upper case I think?

File: src/ui/ui_transaction_nbgl.c [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        content->type = INFO_LONG_PRESS, content->infoLongPress.icon = &C_icon_stellar_64px;
+        content->infoLongPress.text = "Finalize transaction";
+        content->infoLongPress.longPressText = "Hold to confirm";
+    }
+    return true;
+}
+
+static void rejectConfirmation(void) {
+    ui_action_validate_transaction(false);
+    nbgl_useCaseStatus("TRANSACTION\nREJECTED", false, ui_menu_main);
```
Comment (xchapron-ledger) [inline src/ui/ui_transaction_nbgl.c:210]: not upper case

**Review by xchapron-ledger — COMMENTED** [review]
I'm wondering if ui_transaction_nbgl.c could not be simplified if using nbgl_useCaseStaticReview with a callback to retrieve the pairs title and value when needed?
I think that what you did is somehow recreating this mecanism in your app?

**Comment by agrojean-ledger** [comment]
> I'm wondering if ui_transaction_nbgl.c could not be simplified if using nbgl_useCaseStaticReview with a callback to retrieve the pairs title and value when needed? I think that what you did is somehow recreating this mecanism in your app?

I can try. Maybe when I implemented the UI, `nbgl_useCaseStaticReview` wasn't available, I don't remember.

File: Makefile [hunk 4]
```diff
@@ -21,23 +21,27 @@ endif
...
-APP_LOAD_FLAGS=--appFlags 0xa00  # APPLICATION_FLAG_LIBRARY + APPLICATION_FLAG_BOLOS_SETTINGS
+APP_LOAD_PARAMS = --appFlags 0xa00  # APPLICATION_FLAG_LIBRARY + APPLICATION_FLAG_BOLOS_SETTINGS
 endif
-APP_LOAD_PARAMS=$(APP_LOAD_FLAGS) --path "44'/148'" --curve ed25519 $(COMMON_LOAD_PARAMS)
+APP_LOAD_PARAMS += --curve ed25519
+APP_LOAD_PARAMS += --path "44'/148'"
+APP_LOAD_PARAMS +=  $(COMMON_LOAD_PARAMS)
 
 #prepare hsm generation
 ifeq ($(TARGET_NAME),TARGET_NANOS)
```
Comment (tdejoigny-ledger) [inline Makefile:39]: Which icon will be used for nanosp ?

File: src/ui/ui_menu_nbgl.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            break;
+        default:
+            PRINTF("Should not happen !");
+            break;
+    }
+}
+
+static void displaySettingsMenu(void) {
+    nbgl_useCaseSettings("Stellar settings",
+                         0,
```
Comment (tdejoigny-ledger) [inline src/ui/ui_menu_nbgl.c:90]: @better to replace 0, 2 and true by explicit #define

File: src/ui/ui_transaction_nbgl.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        previous_data = G.ui.current_data_index;
+        // Call formatter function.
+        formatter_stack[formatter_index](&G_context.tx_info);
+        PRINTF("Page %d - Item : %s - Value : %s\n",
+               nbPages,
+               G.ui.detail_caption,
+               G.ui.detail_value);
+        // Compute number of lines filled by tag item string.
+        fieldLen = strlen(G.ui.detail_caption);
+        tagItemLineNb = fieldLen / TAG_VAL_LST_ITEM_MAX_CHAR_PER_LINE;
```
Comment (tdejoigny-ledger) [inline src/ui/ui_transaction_nbgl.c:99]: overflow might occur here

File: src/ui/ui_transaction_nbgl.c [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        tagItemLineNb += (fieldLen % TAG_VAL_LST_ITEM_MAX_CHAR_PER_LINE != 0) ? 1 : 0;
+        // Compute number of lines filled by tag value string.
+        fieldLen = strlen(G.ui.detail_value);
+        tagValueLineNb = fieldLen / TAG_VAL_LST_VAL_MAX_CHAR_PER_LINE;
```
Comment (tdejoigny-ledger) [inline src/ui/ui_transaction_nbgl.c:103]: overflow might occur here

File: src/ui/ui_transaction_nbgl.c [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        } else
+        // Otherwise save number of pairs on current page
+        {
+            pagesInfos[nbPages].pagePairNb++;
+        }
+        formatter_index++;
+        continue_loop = (formatter_stack[formatter_index] != NULL) ? true : false;
+    }
+
+    nbPages++;
```
Comment (tdejoigny-ledger) [inline src/ui/ui_transaction_nbgl.c:140]: it seems that `nbPages `is never checked (should not exceed `pagesInfos` size)

**Comment by agrojean-ledger** [comment]
> I'm wondering if ui_transaction_nbgl.c could not be simplified if using nbgl_useCaseStaticReview with a callback to retrieve the pairs title and value when needed? I think that what you did is somehow recreating this mecanism in your app?

As discussed together, using `nbgl_useCaseStaticReview` is not possible because of Tx with multiple operations where some special centered info pages need to be inserted in the flow.

File: Makefile (hunk 4 above)
Comment (agrojean-ledger) [reply Makefile:39]: I'll fix it.

File: src/ui/ui_transaction_nbgl.c [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+               G.ui.detail_caption,
+               G.ui.detail_value);
+        // Compute number of lines filled by tag item string.
+        fieldLen = strlen(G.ui.detail_caption);
+        tagItemLineNb = fieldLen / TAG_VAL_LST_ITEM_MAX_CHAR_PER_LINE;
+        tagItemLineNb += (fieldLen % TAG_VAL_LST_ITEM_MAX_CHAR_PER_LINE != 0) ? 1 : 0;
+        // Compute number of lines filled by tag value string.
+        fieldLen = strlen(G.ui.detail_value);
+        tagValueLineNb = fieldLen / TAG_VAL_LST_VAL_MAX_CHAR_PER_LINE;
+        tagValueLineNb += (fieldLen % TAG_VAL_LST_VAL_MAX_CHAR_PER_LINE != 0) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ui/ui_transaction_nbgl.c:104]: `getNbPagesForTagValueList` depends on other functions such as `getNbTagValuesInPage` and `setNbPairs` copying this code won't make the app simpler or easier to read... I suggest we keep it the way it is for now and we open a discussion regarding exposing proper utils in the SDK.

File: src/ui/ui_transaction_nbgl.c (hunk 8 above)
Comment (agrojean-ledger) [reply src/ui/ui_transaction_nbgl.c]: Fixed.

File: src/ui/ui_transaction_hash_nbgl.c (hunk 2 above)
Comment (agrojean-ledger) [reply src/ui/ui_transaction_hash_nbgl.c:72]: Done.

File: src/ui/ui_menu_nbgl.c (hunk 1 above)
Comment (agrojean-ledger) [reply src/ui/ui_menu_nbgl.c]: Done.

File: src/ui/ui_transaction_nbgl.c (hunk 6 above)
Comment (agrojean-ledger) [reply src/ui/ui_transaction_nbgl.c:107]: How do you suggest we fix it ?

File: src/ui/ui_transaction_nbgl.c (hunk 7 above)
Comment (agrojean-ledger) [reply src/ui/ui_transaction_nbgl.c:111]: How do you suggest we fix it ?

File: .github/workflows/ci-workflow.yml [hunk 10]
```diff
@@ -9,41 +9,19 @@ on:
...
-      - name: Build for ${{ matrix.device.name }}
-        run: |
-          make BOLOS_SDK=${{ matrix.device.path }} DEBUG=1
-
-      - name: Upload app binary
-        uses: actions/upload-artifact@v3
-        with:
-          name: stellar-app-${{ matrix.device.name }}-debug
-          path: bin
+  build_application:
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply .github/workflows/ci-workflow.yml:17]: Added

**Review by xchapron-ledger — DISMISSED** [review]

File: src/ui/ui_menu_nbgl.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    return true;
+}
+
+static void settingsControlsCallback(int token, uint8_t index) {
+    UNUSED(index);
+    switch (token) {
+        case SWITCH_HASH_SET_TOKEN:
+        case SWITCH_SEQUENCE_SET_TOKEN:
+            SETTING_TOGGLE(settings[SETTING_IDX(token)]);
+            break;
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ui/ui_menu_nbgl.c:84]: Sure will do.

**Review by xchapron-ledger — DISMISSED** [review]

**Pull request merged**